### DISCUSSION TITLE
Application crash [Fixed]

### DISCUSSION THREAD

**Discussion by bbday** [cefsharp/CefSharp#3786] [General]
My application crash after a while with last version of cefsharp.offscren v92.0.26 x64 from unmanaged 
libcef.dll exception code: 0x8000003 on window event view i tried to check with .dmp and tell me:

>	**libcef.dll!CefBrowserInfo::NotificationStateLock::NotificationStateLock(CefBrowserInfo * browser_info) Line 568**	C++
 	libcef.dll!CefBrowserInfo::GetMainFrame() Line 230	C++
 	libcef.dll!CefBrowserHostBase::GetFrame(__int64 identifier) Line 637	C++
 	libcef.dll!CefBrowserHostBase::GetMainFrame() Line 626	C++
 	libcef.dll!`anonymous namespace'::browser_get_main_frame(_cef_browser_t * self) Line 253	C++
 	
_loaded browser_info.cc 
https://bitbucket.org/chromiumembedded/cef/src/92ec2dcbdd342d19fbf257009c5b62837c65b72e/libcef/browser/browser_info.cc?at=master_

Disassembly:
```
   565:   // Take the navigation state lock.
   566:   {
   567:     base::AutoLock lock_scope_(browser_info_->notification_lock_);
00007FFBA2A3F575  mov         rcx,r14  
00007FFBA2A3F578  call        base::internal::LockImpl::LockInternalWithTracking (07FFBA1FFBC40h)  
00007FFBA2A3F57D  jmp         CefBrowserInfo::NotificationStateLock::NotificationStateLock+3Ah (07FFBA2A3F4D8h)  
   573: 
   574:   // Take the browser info state lock.
   575:   browser_info_lock_scope_.reset(new base::AutoLock(browser_info_->lock_));
00007FFBA2A3F582  mov         rcx,rbx  
00007FFBA2A3F585  call        base::internal::LockImpl::LockInternalWithTracking (07FFBA1FFBC40h)  
>>> 00007FFBA2A3F58A  jmp         CefBrowserInfo::NotificationStateLock::NotificationStateLock+0BEh (07FFBA2A3F55Ch)  
   568:     CHECK(!browser_info_->notification_state_lock_);
00007FFBA2A3F58C  int         3  
00007FFBA2A3F58D  ud2  
00007FFBA2A3F58F  int         3  
```
 	
How i can fix it? thanks

**Comment by amaitland** [comment]
Load libcef.dll.pdb see https://github.com/cefsharp/CefSharp/wiki/Trouble-Shooting#loading-native-symbols-for-easier-diagnosis for details. 

Make sure you've loaded the CefSharp pdbs as well. It's not clear your stack trace is complete.

**Comment by bbday** [comment]
debug screen app [https://postimg.cc/fJh2rv6H](https://postimg.cc/fJh2rv6H)
debug screen cefsharp [https://postimg.cc/QHhk2g3S](https://postimg.cc/QHhk2g3S)

I loaded release libcef.dll.pdb and i loaded browser_info.cc as debuger ask from cef 92.0.26+g9d87c1f
i can also provide .dmp if you wanna but breakpoint is this, i see on above version (20-08 [https://bitbucket.org/chromiumembedded/cef/src/955097ea774757fe0d08e1c5a566a0c20665095c/libcef/browser/browser_info.cc?at=master](https://bitbucket.org/chromiumembedded/cef/src/955097ea774757fe0d08e1c5a566a0c20665095c/libcef/browser/browser_info.cc?at=master) they remove this point maybe create issue?

I see i have 2 .dmp first is relative about app, second is by cefsharp app, as you can see on second screen it say "TerminateBecauseOutfMemory" is possible throw exception inside application instead crash app?

cefsharp.dmp
[https://1drv.ms/u/s!AgDhCHuE8kYXiVa173nzDT2hbC1e?e=zoE4jX](https://1drv.ms/u/s!AgDhCHuE8kYXiVa173nzDT2hbC1e?e=zoE4jX)

**Reply by amaitland** [reply]
> debug screen app https://postimg.cc/fJh2rv6H
debug screen cefsharp https://postimg.cc/QHhk2g3S

Please provide a stack trace in text format. 

> https://bitbucket.org/chromiumembedded/cef/src/955097ea774757fe0d08e1c5a566a0c20665095c/libcef/browser/browser_info.cc?at=master they remove this point maybe create issue?

Not clear what you mean. You are looking at the master branch, you need to use the 4515 branch. 

> see on second screen it say "TerminateBecauseOutfMemory" is possible throw exception inside application instead crash app?

No that I'm aware of, no. If you are using a 32bit app make sure it's LargeAddressAware.

**Reply by bbday** [reply]
do you mean this?

```
        libcef.dll!CefBrowserInfo::NotificationStateLock::NotificationStateLock(CefBrowserInfo * browser_info) Line 568	C++
 	libcef.dll!CefBrowserInfo::GetMainFrame() Line 230	C++
 	libcef.dll!CefBrowserHostBase::GetFrame(__int64 identifier) Line 637	C++
 	libcef.dll!CefBrowserHostBase::GetMainFrame() Line 626	C++
 	libcef.dll!`anonymous namespace'::browser_get_main_frame(_cef_browser_t * self) Line 253	C++
	CefSharp.Core.Runtime.dll!CefBrowserCToCpp::GetMainFrame(void)	Unknown
 	[Managed to Native Transition]	
 	[Frames below may be incorrect and/or missing]	
 	[Frames below may be incorrect and/or missing]	
 	[Frames below may be incorrect and/or missing, no binaries loaded for mylibrary.dll]	
 	[Frames below may be incorrect and/or missing]	
 	[Native to Managed Transition]	
 	CefSharp.Core.Runtime.dll!CefCToCppRefCounted<class CefRequestCallbackCToCpp,class CefRequestCallback,struct _cef_request_callback_t>::Wrap(struct _cef_request_callback_t *)	Unknown
 	libcef.dll!CefResourceRequestHandlerCToCpp::OnResourceLoadComplete(scoped_refptr<CefBrowser> browser={...}, scoped_refptr<CefFrame> frame={...}, scoped_refptr<CefRequest> request={...}, scoped_refptr<CefResponse> response={...}, <unnamed-tag> status=UR_SUCCESS, __int64 received_content_length=748) Line 233	C++
 	libcef.dll!net_service::`anonymous namespace'::InterceptedRequestHandlerWrapper::CallHandlerOnComplete(net_service::`anonymous namespace'::InterceptedRequestHandlerWrapper::RequestState * state, const network::URLLoaderCompletionStatus & status) Line 1023	C++
 	libcef.dll!net_service::`anonymous namespace'::InterceptedRequestHandlerWrapper::OnRequestComplete(int request_id=73275, const network::ResourceRequest & request, const network::URLLoaderCompletionStatus & status) Line 981	C++
 	libcef.dll!net_service::InterceptedRequest::OnDestroy() Line 1051	C++
 	[Inline Frame] libcef.dll!base::OnceCallback<void (unsigned int, const std::__1::string &)>::Run(unsigned int args, const std::__1::basic_string<char,std::__1::char_traits<char>,std::__1::allocator<char>> & args) Line 98	C++
 	libcef.dll!mojo::InterfaceEndpointClient::NotifyError(const absl::optional<mojo::DisconnectReason> & reason) Line 687	C++
 	libcef.dll!mojo::internal::MultiplexRouter::ProcessNotifyErrorTask(mojo::internal::MultiplexRouter::Task * task, mojo::internal::MultiplexRouter::ClientCallBehavior client_call_behavior, base::SequencedTaskRunner * current_task_runner) Line 997	C++
 	libcef.dll!mojo::internal::MultiplexRouter::ProcessTasks(mojo::internal::MultiplexRouter::ClientCallBehavior client_call_behavior=ALLOW_DIRECT_CLIENT_CALLS, base::SequencedTaskRunner * current_task_runner=0x00000462002613e0) Line 908	C++
 	libcef.dll!mojo::internal::MultiplexRouter::OnPipeConnectionError(bool force_async_dispatch) Line 820	C++
 	[Inline Frame] libcef.dll!base::OnceCallback<void ()>::Run() Line 98	C++
 	libcef.dll!mojo::Connector::HandleError(bool force_pipe_reset, bool force_async_handler) Line 665	C++
 	[Inline Frame] libcef.dll!base::RepeatingCallback<void (unsigned int, const mojo::HandleSignalsState &)>::Run(unsigned int args=9, const mojo::HandleSignalsState & args={...}) Line 166	C++
 	libcef.dll!mojo::SimpleWatcher::OnHandleReady(int watch_id, unsigned int result=9, const mojo::HandleSignalsState & state={...}) Line 279	C++
 	libcef.dll!mojo::SimpleWatcher::Context::Notify(unsigned int result=9, MojoHandleSignalsState signals_state, unsigned int flags) Line 94	C++
 	libcef.dll!mojo::SimpleWatcher::Context::CallNotify(const MojoTrapEvent * event=0x0000000a25efed70) Line 63	C++
 	libcef.dll!mojo::core::WatcherDispatcher::InvokeWatchCallback(unsigned __int64 context, unsigned int result=9, const mojo::core::HandleSignalsState & state, unsigned int flags=0) Line 95	C++
 	libcef.dll!mojo::core::Watch::InvokeCallback(unsigned int result=9, const mojo::core::HandleSignalsState & state={...}, unsigned int flags=0) Line 79	C++
 	libcef.dll!mojo::core::RequestContext::~RequestContext() Line 73	C++
 	libcef.dll!mojo::core::NodeChannel::OnChannelMessage(const void * payload, unsigned __int64 payload_size, std::__1::vector<mojo::PlatformHandle,std::__1::allocator<mojo::PlatformHandle>> handles) Line 828	C++
 	libcef.dll!mojo::core::Channel::TryDispatchMessage(base::span<const char,18446744073709551615> buffer, unsigned __int64 * size_hint=0x0000000a25eff5a0) Line 926	C++
 	libcef.dll!mojo::core::Channel::OnReadComplete(unsigned __int64 bytes_read, unsigned __int64 * next_read_size_hint=0x0000000a25eff5a0) Line 826	C++
 	[Inline Frame] libcef.dll!mojo::core::`anonymous namespace'::ChannelWin::OnReadDone(unsigned __int64 bytes_read) Line 296	C++
 	libcef.dll!mojo::core::`anonymous namespace'::ChannelWin::OnIOCompleted(base::MessagePumpForIO::IOContext * context, unsigned long bytes_transfered=56, unsigned long error) Line 282	C++
 	libcef.dll!base::MessagePumpForIO::WaitForIOCompletion(unsigned long timeout) Line 797	C++
 	libcef.dll!base::MessagePumpForIO::WaitForWork(base::MessagePump::Delegate::NextWorkInfo next_work_info) Line 770	C++
 	libcef.dll!base::MessagePumpForIO::DoRunLoop() Line 752	C++
 	libcef.dll!base::MessagePumpWin::Run(base::MessagePump::Delegate * delegate) Line 80	C++
 	libcef.dll!base::sequence_manager::internal::ThreadControllerWithMessagePumpImpl::Run(bool application_tasks_allowed, base::TimeDelta timeout) Line 471	C++
 	libcef.dll!base::RunLoop::Run(const base::Location & location) Line 136	C++
 	libcef.dll!base::Thread::Run(base::RunLoop * run_loop) Line 325	C++
 	libcef.dll!content::BrowserProcessIOThread::IOThreadRun(base::RunLoop * run_loop=0x0000000a25effa70) Line 128	C++
 	libcef.dll!base::Thread::ThreadMain() Line 399	C++
 	libcef.dll!base::`anonymous namespace'::ThreadFunc(void * params=0x0000046200270690) Line 113	C++
 	kernel32.dll!BaseThreadInitThunk()	Unknown
 	ntdll.dll!RtlUserThreadStart()	Unknown
```

cause i see OnResourceLoadComplete i add my class maybe can be the issue:

    class InterceptorHandler : ResourceRequestHandler
    {
        private readonly ConcurrentDictionary<ulong, MemoryStream> _dicReq;

        public delegate void ResponseProcessHandlern(string data, string url);
        public event ResponseProcessHandlern OnResponseProcessData;

        public InterceptorHandler(ConcurrentDictionary<ulong, MemoryStream> dicReq) : base()
        {
            _dicReq = dicReq;
        }

        protected override IResponseFilter GetResourceResponseFilter(IWebBrowser chromiumWebBrowser, IBrowser browser, IFrame frame, IRequest request, IResponse response)
        {
            bool ignoreResource = request.ResourceType == ResourceType.FontResource || request.ResourceType == ResourceType.Favicon || request.ResourceType == ResourceType.Stylesheet ||
                request.ResourceType == ResourceType.Image && string.IsNullOrEmpty(new Uri(request.Url).Query);

            if (ignoreResource)
                return null;

            var memory = new MemoryStream();
            _dicReq.TryAdd(request.Identifier, memory);
            return new CefSharp.ResponseFilter.StreamResponseFilter(memory);
        }

        protected override void OnResourceLoadComplete(IWebBrowser chromiumWebBrowser, IBrowser browser, IFrame frame, IRequest request, IResponse response, UrlRequestStatus status, long receivedContentLength)
        {
            if (_dicReq.TryRemove(request.Identifier, out var stream))
            {
                try
                {
                    if (stream != null && stream.CanRead)
                    {
                        var bytes = stream.ToArray();
                        OnResponseProcessData?.Invoke(response.Charset == "utf-8" ? Encoding.UTF8.GetString(bytes) : Encoding.Default.GetString(bytes), request.Url);             
                    }
                }
                catch (Exception ex)
                {
                    
                }
                finally
                {
                    try
                    {
                        if (stream != null && stream.CanRead)
                            stream.Dispose();
                    }
                    catch { }
                }
            }
        }

        protected override CefReturnValue OnBeforeResourceLoad(IWebBrowser chromiumWebBrowser, IBrowser browser, IFrame frame, IRequest request, IRequestCallback callback)
        {
            try
            {
                if (Uri.TryCreate(request.Url, UriKind.Absolute, out _) == false)
                {
                    //If we're unable to parse the Uri then cancel the request
                    // avoid throwing any exceptions here as we're being called by unmanaged code
                    return CefReturnValue.Cancel;
                }

                //NOTE: When executing the callback in an async fashion need to check to see if it's disposed
                if (!callback.IsDisposed)
                {
                    using (callback)
                    {
                        if (string.IsNullOrEmpty(request.GetHeaderByName"sec-ch-ua")))
                            request.SetHeaderByName("sec-ch-ua", "\"Chromium\";v=\"92\", \" Not A;Brand\";v=\"99\", \"Google Chrome\";v=\"92\"", true);
                    }
                }

                return CefReturnValue.Continue;
            }
            catch (Exception ex)
            {
                return CefReturnValue.Cancel;
            }
        }
    }

**Reply by amaitland** [reply]
> do you mean this?

Yes thank you. 

Please load the symbols for mylibrary.dll and provide an updated stack trace.

**Comment by bbday** [comment]
Thanks for support, look like i fixed issue, after i loaded all refercense issue was call this function from OnResponseProcessData event.

```
OnResponseProcessData += Process

 private void Process(string message, string url)
{
   **if(!browser.IsDisposed && browser.IsBrowserInitialized && browser.CanExecuteJavascriptInMainFrame)**
       browser.ExecuteScriptAsync(operation);
}
```

**Reply by amaitland** [reply]
The underlying crash should hopefully be resolved in the next major release. For reference https://magpcss.org/ceforum/viewtopic.php?f=6&t=18631#p50122
